### PR TITLE
fix: validate query populate

### PR DESCRIPTION
**Pull request by innerdvations**
### What does it do?

Query populate parameters are now validated instead of only sanitizing, meaning that a query populate param that was invalid (such as targeting a private relation) was working with the param silently removed will now throw an error instead.

This applies to the validator util itself, along with the admin API and the content API

Also:

- traversals (such as travereQueryPopulate) should not modify the thing they are traversing, so the wildcard populate expansion has been removed from there and made into a visitor that is run by the sanitizer and validator
- full paths are now sent with error message

### Why is it needed?

In v4, validation was added for security but was a somewhat breaking change. For populate query params, there was no direct security risk involved, so the breaking change wasn't made.

However, it makes user queries harder to debug and is bad practice in general, so we always intended to make this breaking change in v5.

### How to test it?

Send a populate query with invalid populate parameters (such as targeting a private field or one you don't have permission for) -- 
Without this PR it should have allowed the query to work but ignored your populate.
With this PR, you will receive an error (a ValidationError, to prevent potentially leaking information about private parts of schemas like with a ForbiddenError)

Existing tests have been updated to match the new expected behaviour

### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
Cc @Boegie19 here you go 😁

@innerdvations shouldn't we throw a 403 here since the user doesn't have the correct permissions?

**Comment by innerdvations** [comment]
> @innerdvations shouldn't we throw a 403 here since the user doesn't have the correct permissions?

Yep, fixed that in the description 😆

**Comment by Boegie19** [comment]
@innerdvations Could you add the name of the field/relation you are trying to populate in the error?

This will make debuging better since then you also get the first field/relation that is failing

**Comment by innerdvations** [comment]
> @innerdvations Could you add the name of the field/relation you are trying to populate in the error?
> 
> This will make debuging better since then you also get the first field/relation that is failing since that would be an other W for dev debug experience would even be better if it showed the full path to it but that would most likley be to much work Also this could be done later as this would not be breaking
> 
> still this PR by itself is already a masive W

Yep, would love to follow up with improved error messages! But that will probably have to come after v5 RC

**Comment by innerdvations** [comment]
@Boegie19 We've added the path to all errors thrown in this PR :)

File: packages/core/admin/server/src/services/permission/permissions-manager/validate.ts [hunk 1]
```diff
@@ -109,7 +124,7 @@ export default ({ action, ability, model }: any) => {
       // Remove fields hidden from the admin
       traverseEntity(throwHiddenFields, { schema }),
       // Remove not allowed fields (RBAC)
-      // @ts-expect-error lodash types
+      // @ts-ignore-error lodash types
```
Comment (innerdvations) [inline packages/core/admin/server/src/services/permission/permissions-manager/validate.ts:127]:
```suggestion
      // @ts-expect-error lodash types
```

**Comment by Boegie19** [comment]
Thanks this will he so helpfull

**Comment by innerdvations** [comment]
> > @innerdvations shouldn't we throw a 403 here since the user doesn't have the correct permissions?
> 
> Yep, fixed that in the description 😆

Sorry, looking at again, no, it will be a 400 to prevent leaking data -- if the user doesn't have access to an attribute, the user shouldn't even know that attribute exists, even though in this case it's a bit weird that you see the field itself as an ID but we would still be revealing that it is a relation vs a number. In the future we can review and maybe add more specific error codes to the various validations for cases where we know for sure it's not leaking anything.

**Comment by derrickmehaffy** [comment]
> > > @innerdvations shouldn't we throw a 403 here since the user doesn't have the correct permissions?
> > 
> > Yep, fixed that in the description 😆
> 
> Sorry, looking at again, no, it will be a 400 to prevent leaking data -- if the user doesn't have access to an attribute, the user shouldn't even know that attribute exists, even though in this case it's a bit weird that you see the field itself as an ID but we would still be revealing that it is a relation vs a number. In the future we can review and maybe add more specific error codes to the various validations for cases where we know for sure it's not leaking anything.

I think this is more "security through obscurity" and isn't really leaking data. Other larger API like say GitHub do the same. So long as the security is there, it's fine to show the error.

**Comment by innerdvations** [comment]
> > > > @innerdvations shouldn't we throw a 403 here since the user doesn't have the correct permissions?
> > > 
> > > 
> > > Yep, fixed that in the description 😆
> > 
> > 
> > Sorry, looking at again, no, it will be a 400 to prevent leaking data -- if the user doesn't have access to an attribute, the user shouldn't even know that attribute exists, even though in this case it's a bit weird that you see the field itself as an ID but we would still be revealing that it is a relation vs a number. In the future we can review and maybe add more specific error codes to the various validations for cases where we know for sure it's not leaking anything.
> 
> I think this is more "security through obscurity" and isn't really leaking data. Other larger API like say GitHub do the same. So long as the security is there, it's fine to show the error.

Not exactly, we try to consider non-public schema attributes private, although it's admittedly low severity if it happens. This is less like github documenting what exists (even if you don't have permissions for it) and more like if github had columns in their db like "charge_this_client_extra" that were only supposed to be seen internally. It's less about the security of making it easier or harder to hack into that data, and more about not revealing the data itself even existing.

**Comment by derrickmehaffy** [comment]
🤔 I suppose yes, still not a huge fan but if it's what we have time for now and can do then not much else to add.

My main thing is we fall back on overusing 400 wayyyyyy too much and we rarely use the proper http status codes like we should. 400 should be our absolutely last option when absolutely nothing else fits.

**Comment by innerdvations** [comment]
> 🤔 I suppose yes, still not a huge fan but if it's what we have time for now and can do then not much else to add.
> 
> My main thing is we fall back on overusing 400 wayyyyyy too much and we rarely use the proper http status codes like we should. 400 should be our absolutely last option when absolutely nothing else fits.
> 
> 5xx is the same honestly we pretty much only use 500 and no other 5xx code.

It's true that this case (populate) is one area where a 403 might make sense, since they do know the field exists. I will do a follow-up once I have a chance to think about it and adjust the error thrown if I don't see any potential issues with it :)

If you can keep a list of other specific things that throw the wrong error code (not necessarily in validation) that would help, maybe we could tackle them all at once

**Review by Convly — APPROVED** [review]
LGTM, would be nice to haver another dev QAing this one

**Comment by innerdvations** [comment]
Just discovered an issue with nested populates this mornning that I need to investigate and add a test for, so back to draft. If it makes sense I'll update the status thrown here too.

**Pull request merged**